core/src/main/kotlin/org/evomaster/core/output/clustering/DBSCANClusterer.kt

```diff
@@ -0,0 +1,137 @@
+ * - Instantiate (see [Clusterer] for example).
+ * - Invoke [performClustering]
+ *
+ * The parameters:
+ * [V] - the type of objects to be clustered.
+ * [minimumNumberOfClusterMembers] - DBSCAN filters out outliers, so it does not allow single
+ * element clusters. The default value for this is [minimumNumberOfClusterMembers = 2], since
+ * we want to have the possibility of small clusters, where appropriate.
+ * [metric] - the type of metric being used. Must extend the [DistanceMetric] abstract class.
+ * [inputValues] - the collection of objects to be clustered.
+ * [epsilon] - the density neighbourhood of an element. To cluster error messages, we use a
+ * normalized distance (see [DistanceMetricAction]). The default value is 0.5 - that is, if
+ * half of the error message is the same, the Actions will be clustered together. The idea is
+ * to allow messages that differ in the details (timestamps, object references, host and port,
+ * etc.) but are otherwise similar to cluster together.
+ *
+ */
+
+class DBSCANClusterer<V> {
+    private var epsilon: Double = 1.0
+    private var minimumNumberOfClusterMembers: Int = 2
+    private lateinit var metric: DistanceMetric<V>
+    private lateinit var inputValues: List<V>
+    private var visitedPoints: HashSet<V> = HashSet()
+
+    constructor(values: Collection<V> , epsilon: Double = 0.5, minimumMembers: Int = 2, metric: DistanceMetric<V>){
+        setEpsilon(epsilon)
+        setMinimumNrOfMembers(members = minimumMembers)
+        setInputValues(values)
+        setMetric(metric)
+    }
+
+    fun setMinimumNrOfMembers(members: Int){
+        if(members > 1) this.minimumNumberOfClusterMembers = members
+    }
+
+    fun setEpsilon(epsilon: Double){
+        if(epsilon > 0.0) this.epsilon = epsilon
+    }
+
+    fun setMetric(metric: DistanceMetric<V>){
+        this.metric = metric
+    }
+
+    fun setInputValues(values: Collection<V>){
+        if(values.isEmpty()){
+            throw Exception("DBSCAN: The list of values appears to be empty")
+        }
+        this.inputValues = values.toList()
+    }
+
+    private fun getNeighbours(inputValue: V): MutableList<V>{
+        val neighbours: MutableList<V> = mutableListOf()
+        for (i in 0..inputValues.size-1){
+            val candidate = inputValues.get(i)
+            if(metric.calculateDistance(inputValue, candidate) <= epsilon){
+                neighbours.add(candidate)
+            }
+        }
+        return neighbours
+    }
+
+    private fun mergeRightToLeft(neighbours1: MutableList<V>, neighbours2: MutableList<V>): MutableList<V>{
+        for ( i in 0..neighbours2.size-1){
+            val temPt = neighbours2.get(i)
+            if(!neighbours1.contains(temPt)) neighbours1.add(temPt)
+        }
+        return neighbours1
+    }
+
+
+    fun performCLustering(): MutableList<MutableList<V>>{
+        if(!::inputValues.isInitialized) throw Exception("DBSCAN: List of inputs has not been initialized")
```

Comment:
rather throw an `IllegalStateException`. Please check what I wrote in `for_developers.md`, in particular the section on `PRE AND POST CONDITIONS`.